packages/dev/core/src/Collisions/gpuPicker.ts

```diff
@@ -0,0 +1,314 @@
+            }
+
+            const r = (id & 0xff0000) >> 16;
+            const g = (id & 0x00ff00) >> 8;
+            const b = (id & 0x0000ff) >> 0;
+            this._idMap[`${r}_${g}_${b}`] = index;
+            id++;
+
+            if (mesh.hasInstances) {
+                const instances = (mesh as Mesh).instances;
+                const colorData = new Float32Array(4 * (instances.length + 1));
+                const engine = mesh.getEngine();
+
+                colorData[0] = r / 255.0;
+                colorData[1] = g / 255.0;
+                colorData[2] = b / 255.0;
+                colorData[3] = 1.0;
+                for (let i = 0; i < instances.length; i++) {
+                    const instance = instances[i];
+                    const r = (id & 0xff0000) >> 16;
+                    const g = (id & 0x00ff00) >> 8;
+                    const b = (id & 0x0000ff) >> 0;
+                    this._idMap[`${r}_${g}_${b}`] = this._pickableMeshes.indexOf(instance);
+
+                    colorData[(i + 1) * 4] = r / 255.0;
+                    colorData[(i + 1) * 4 + 1] = g / 255.0;
+                    colorData[(i + 1) * 4 + 2] = b / 255.0;
+                    colorData[(i + 1) * 4 + 3] = 1.0;
+                    id++;
+                }
+
+                const buffer = new VertexBuffer(engine, colorData, VertexBuffer.ColorKind, false, false, 4, true);
+                (mesh as Mesh).setVerticesBuffer(buffer, true);
+            } else {
+                this._idColors[mesh.uniqueId] = Color3.FromInts(r, g, b);
+            }
+        }
+    }
+
+    /**
+     * Execute a picking operation
+     * @param x defines the X coordinates where to run the pick
+     * @param y defines the Y coordinates where to run the pick
+     * @param scene defines the scene to pick from
+     * @param disposeWhenDone defines a boolean indicating we do not want to keep resources alive
+     * @returns A promise with the picking results
+     */
+    public pickAsync(x: number, y: number, scene: Scene, disposeWhenDone = true): Promise<Nullable<AbstractMesh>> {
+        const engine = scene.getEngine();
+        const rttSizeW = engine.getRenderWidth();
+        const rttSizeH = engine.getRenderHeight();
+
+        if (!this._readbuffer) {
+            this._readbuffer = new Uint8Array(engine.isWebGPU ? 256 : 4); // Because of block alignment in WebGPU
+        }
+
+        this._meshRenderingCount = 0;
+        // Ensure ints
+        x = x >> 0;
+        y = y >> 0;
+
+        if (x < 0 || y < 0 || x >= rttSizeW || y >= rttSizeH) {
+            return Promise.resolve(null);
+        }
+
+        // Invert Y
+        y = rttSizeH - y;
+
+        if (!this._pickingTexure) {
+            this._createRenderTarget(scene, rttSizeW, rttSizeH);
+        } else {
+            const size = this._pickingTexure.getSize();
+
+            if (size.width !== rttSizeW || size.height !== rttSizeH || this._cachedScene !== scene) {

```

Comment:
Maybe for future optimizations:

Let the user define a `factor` (in the constructor?) so that the render texture is created with smaller dimensions than the screen (`texture_dimensions = screen_size / factor`). Often, we don't need picking to be pixel perfect, so we could get some perf gain + memory savings by rendering in a smaller texture.